### PR TITLE
Add support for more metadata in UVCal, add metadata_only support

### LINKED ISSUE #735: Add telescope location information to UVCal
Right now, when looking at calibration solutions, I need to also read in a uvdata object just to get the telescope location so I can convert the times to LST. It would make things more efficient to only need to read in uvcal objects.

### PR DESCRIPTION
**Pull request by bhazelton**
## Description

Add support for telescope_location, antenna_positions and lst_array in UVCal object and file types. Add support for a metadata_only mode in UVCal, including options to only read the metadata when reading in calibration files.

## Motivation and Context

This is preparing for creating a function to initialize a UVCal object from a UVData object (#753)
closes #735 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Breaking change checklist:
- [x] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to reflect my changes (if appropriate).
- [x] My change includes backwards compatibility and deprecation warnings (if possible).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

### CODE REVIEW

**Review by dstorer — APPROVED** [review]
This looks good to me - thanks for implementing this!

**Comment by bhazelton** [comment]
@mkolopanis I'm afraid this PR adds more deprecation warnings which is upsetting hera_qm tests.

**Comment by bhazelton** [comment]
@jsdillon This PR is starting a process to begin requiring some additional metadata on UVCal objects: telescope_location, lst_array, and antenna_positions. There have been many requests for these to exist on UVCal objects from users, mostly HERA users.

I'm working on a function to transfer metadata (including those) from a UVData object, but wanted to give you an early heads up about this change. If you're concerned about it in any way please let us know!

**Comment by jsdillon** [comment]
Thanks for the heads up @bhazelton. Are these extra parameters going to be required on read, or just on write? We've got a lot of calibration solution files sitting around that don't have these and it would be a real pain to have to update them all or rewrite code that reads them. 

I'd be happy if, on read, missing parameters produces a warning, and on write, they produce an error.

**Comment by bhazelton** [comment]
@jsdillon the telescope_location and lst_array will be totally fine because as long as the telescope_name is in our known_telescopes (and HERA is), those can be set by the reader on the object.

The antenna_positions are trickier. I was thinking we would want to make that a required parameter eventually (although the time scale is negotiable) in which case it would error if you run a check on the object. We can always make it not error on the read by setting the `run_check` keyword to False, but then none of the checks are run. I suppose we could add another option that was specific to that parameter. Or we could decide to only make that parameter optional or put transitioning it to required on a longer time scale than the others. What do you think?

**Comment by jsdillon** [comment]
For telescope_location, the one issue I might imagine is with the RIMEz simulations, for example those used in HERA validation. Perhaps @zacharymartinot can work with you to include RIMEz in pyuvdata appropriately?

For the antenna locations, this is tricky. I'd love for all UVCal objects to have it, but backwards compatibility in pipelines is a concern (and I'd be hesitant to turn off checking in any pipeline steps). Is it possible to have a parameter that is required for writing but not for reading? Otherwise, I suspect we'll want a long deprecation period.

**Comment by bhazelton** [comment]
@jsdillon there is not support currently for something that is required on write but not on read in the structure of pyuvdata objects. It's a bit strange to have something like that because then you could read it in but not write it back out without adding the antenna_positions. Is that really a useful mechanism?

**Comment by jsdillon** [comment]
If it existed, it might have helped, but I don't think it's worth building. 

A long warning period is probably necessary.

**Comment by bhazelton** [comment]
@jsdillon We came up with a better solution in our meeting this morning. We will add antenna positions as optional metadata to our known telescopes list. I will add the HERA positions and then if they are missing we can add them to the object using that list. That will make sure they are defined on the object (at least for HERA data) even if they are missing in the file that is read in.

**Comment by jsdillon** [comment]
Good idea! That sounds fine to me.

**Review by theo-kunicki — APPROVED** [review]
I've looked over this, and it seems like these changes will be very useful -- thank you, Bryna! I'll let you know if I run into any problems once it's been merged, but I can't see anything obvious right now.

**Review by plaplant — COMMENTED** [review]
Thanks so much @bhazelton! This is a tremendous amount of work, and the overhaul of UVCal tests was long-overdue. I have a few minor comments, but generally everything looks excellent.

**Comment by plaplant** [comment]
Also we definitely want to make sure we update the changelog.

File: pyuvdata/uvcal/calfits.py [hunk 1]
```diff
@@ -126,6 +126,13 @@ def write_calfits(
         # Conforming to fits format
         prihdr["SIMPLE"] = True
         prihdr["TELESCOP"] = self.telescope_name
+        if self.telescope_location is not None:
+            prihdr["ARRAYX"] = self.telescope_location[0]
+            prihdr["ARRAYY"] = self.telescope_location[1]
+            prihdr["ARRAYZ"] = self.telescope_location[2]
+            prihdr["LAT     "] = self.telescope_location_lat_lon_alt_degrees[0]
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvcal/calfits.py:133]: I think it doesn't matter and I'm happy to change it. The FITS format forces them all to be length 8 and adds padding if they're not that long.

File: pyuvdata/uvdata/fhd.py [hunk 2]
```diff
@@ -343,174 +561,37 @@ def read_fhd(
...
-                keyword = field
-                if len(keyword) > 8:
-                    keyword = field.replace("_", "")
-
-                value = layout[field][0]
-                if isinstance(value, bytes):
-                    value = value.decode("utf8")
-
-                self.extra_keywords[keyword.upper()] = value
+            (
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata/fhd.py:570]: Yeah, that's a good idea. I'll make that change.

File: pyuvdata/uvcal/fhd_cal.py [hunk 3]
```diff
@@ -62,159 +74,126 @@ def read_fhd_cal(
...
-            self.jones_array = np.array([-5])
-        else:
-            self.jones_array = np.array([-5, -6])
+        # this is generated in FHD by subtracting the JD of neighboring
+        # integrations. This can have limited accuracy, so it can be slightly
+        # off the actual value.
+        # (e.g. 1.999426... rather than 2)
+        time_res = obs_data["TIME_RES"][0]
+        # time_res is constrained to be a scalar currently
+        self.integration_time = np.float(time_res)
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvcal/fhd_cal.py:104]: Sure, happy to change it.

File: pyuvdata/uvcal/fhd_cal.py [hunk 4]
```diff
@@ -62,159 +74,126 @@ def read_fhd_cal(
...
         self.spw_array = np.array([0])
 
-        self.Nfreqs = int(cal_data["n_freq"][0])
-        self.freq_array = np.zeros(
-            (self.Nspws, len(cal_data["freq"][0])), dtype=np.float_
-        )
-        self.freq_array[0, :] = cal_data["freq"][0]
-        self.channel_width = float(np.mean(np.diff(self.freq_array)))
+        self.Nfreqs = int(obs_data["N_FREQ"][0])
+        self.freq_array = np.zeros((1, len(bl_info["FREQ"][0])), dtype=np.float_)
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvcal/fhd_cal.py:85]: Sure, happy to change it.

**Comment by bhazelton** [comment]
Thanks for the review @plaplant!! I know it was a lot of changes. I think I addressed all your comments and I updated the changelog. Can you take a look at the changelog and make sure it all makes sense? There were a lot of changes.

**Review by plaplant — APPROVED** [review]
Thanks for making the changes @bhazelton. I think the changelog updates make sense. Feel free to merge anytime. We should also update the hera_qm tests after this has been merged.

**Pull request merged**